### PR TITLE
IBP-4131-DuplicateKeyInGermplasmListDataMap

### PR DESCRIPTION
**Pull request by gelli-leafnode**
If list has duplicate entry (gid), return previous germplasmlistdata

### CODE REVIEW

File: src/main/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImpl.java [hunk 1]
```diff
@@ -104,7 +104,7 @@ public StudyEntryDto replaceStudyEntry(final Integer studyId, final Integer entr
 			germplasmList.getListData().stream().map(l -> mapper.map(l, StudyEntryDto.class)).collect(Collectors.toList());
 
 		final Map<Integer, GermplasmListData> germplasmListDataMap =
-			germplasmList.getListData().stream().collect(Collectors.toMap(GermplasmListData::getGermplasmId, g -> g));
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImpl.java:107]: I'm not sure if we should be using GID as map key here given we can have duplicate GIDs within a list. I think we should be using entry number (entry ID) as key. 
What do you think @mderamos ?

Comment (mderamos) [reply src/main/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImpl.java:107]: Hi @darla-leafnode I agree. We should use the  entry number. Sorry, I missed it. 
Hi @gelli-leafnode , please revise the code to  use entry number as key. Thanks!

File: src/test/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImplTest.java [hunk 2]
```diff
@@ -188,4 +192,71 @@ public void testGetEntryDescriptorColumns() {
...
+		datasetDTO.setDatasetTypeId(DatasetTypeEnum.PLOT_DATA.getId());
+		final List<DatasetDTO> datasetDTOS = Collections.singletonList(datasetDTO);
+		Mockito.when(this.datasetService.getDatasets(studyId, new HashSet<>(Arrays.asList(DatasetTypeEnum.PLOT_DATA.getId()))))
+			.thenReturn(datasetDTOS);
+
+		try {
+			final List<StudyEntryDto> studyEntryDtos = this.studyEntryService.createStudyEntries(studyId, germplasmListId);
+			Assert.notNull(studyEntryDtos, "Duplicate gid in list should be accepted. ");
+		} catch (final Exception e) {
+			e.printStackTrace();
```
Comment (darla-leafnode) [inline src/test/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImplTest.java:220]: Remove` e.printStackTrace()`

File: src/test/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImplTest.java [hunk 3]
```diff
@@ -188,4 +192,71 @@ public void testGetEntryDescriptorColumns() {
...
+		final DatasetDTO datasetDTO = new DatasetDTO();
+		datasetDTO.setDatasetId(datasetId);
+		datasetDTO.setDatasetTypeId(DatasetTypeEnum.PLOT_DATA.getId());
+		final List<DatasetDTO> datasetDTOS = Collections.singletonList(datasetDTO);
+		Mockito.when(this.datasetService.getDatasets(studyId, new HashSet<>(Arrays.asList(DatasetTypeEnum.PLOT_DATA.getId()))))
+			.thenReturn(datasetDTOS);
+
+		try {
+			final List<StudyEntryDto> studyEntryDtos = this.studyEntryService.createStudyEntries(studyId, germplasmListId);
+			Assert.notNull(studyEntryDtos, "Duplicate gid in list should be accepted. ");
```
Comment (darla-leafnode) [inline src/test/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImplTest.java:218]: Make assertion on count of study entry list

**Review by darla-leafnode — APPROVED** [review]

File: src/test/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImplTest.java [hunk 4]
```diff
@@ -188,4 +192,81 @@ public void testGetEntryDescriptorColumns() {
...
+		Mockito.when(this.datasetService.getDatasets(studyId, new HashSet<>(Arrays.asList(DatasetTypeEnum.PLOT_DATA.getId()))))
+			.thenReturn(datasetDTOS);
+		Mockito.when(this.middlewareStudyEntryService.saveStudyEntries(ArgumentMatchers.eq(studyId), ArgumentMatchers.anyList())).thenReturn(entryDtos);
+
+		try {
+			final List<StudyEntryDto> studyEntryDtos = this.studyEntryService.createStudyEntries(studyId, germplasmListId);
+			Assert.notNull(studyEntryDtos, "Duplicate gid in list should be accepted. ");
+			org.junit.Assert.assertEquals("Must return same germplasm list data count", listData.size(),studyEntryDtos.size());
+		} catch (final Exception e) {
+			Assert.isNull(e, "Duplicate gid in list should be accepted, no exception");
```
Comment (darla-leafnode) [inline src/test/java/org/ibp/api/java/impl/middleware/study/StudyEntryServiceImplTest.java:231]: Use Assert.fail inside catch clause

**Pull request merged**